mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -0,0 +1,602 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! This module creates a faucet "worker". It consists of a worker thread,
+//! and a handle to the worker thread, which can be used to asynchronously get
+//! the output of the worker.
+//!
+//! The worker ensures that the faucet is ready for operation by making self-
+//! payments that split TxOut's if the number of pre-split TxOut's falls below a
+//! threshold. TxOut's that are ready to be used get added to a tokio queue.
+//!
+//! Other threads may dequeue UTXOs from such a queue and then use them.
+//! The queue also carries a one-shot channel which can give the worker a
+//! SubmitTxResponse, or signal that an error occurred. The worker then follows
+//! up on what happens with the UTXO, and eventually requeues it if the
+//! transaction doesn't resolve successfully.
+//!
+//! The worker uses its own thread, but uses async-friendly tokio primitives.
+//! The worker does not require to be launched from the context of a tokio
+//! runtime.
```

Comment:
It kind of seems like we're doing a custom implementation of futures and an executor here? This is pretty impressive what you made here - but a lot of it also seems like it could be done using Rust's normal async primitives + an executor like Tokio/Block-on/Async-std.

We're avoiding using the Tokio executor - but why exactly? It seems like there's a boilerplate could be avoided by functionality already existing in Tokio and other executors like spawned tasks, threads, and a ton of helper structs that provide things like queues that can work in an async context:
https://docs.rs/futures/0.3.21/futures/future/index.html
https://docs.rs/futures/0.3.21/futures/stream/index.html

I.e. like these for qeueing up tasks:
https://docs.rs/futures/0.3.21/futures/stream/struct.FuturesUnordered.html
or 
https://docs.rs/futures/0.3.21/futures/stream/struct.FuturesOrdered.html

I think we lose some level of maintaibility here, ability to work with native rust concurrency, and also some pre-done optimization that other executors have. 

If this already works and the faucet is ready to go, we can/should go with this, but it could in the future potentially we could go towards a more "native async" approach.

Reply:
I don't think I made a custom implementation of futures or an executor here -- I think all I did was create a background worker thread. 

I considered doing it another way:
* Replace `std::thread::spawn` with `tokio::task::spawn`
* Replace `std::thread::sleep` with `tokio::task::sleep(..).await`.

The main reason that I didn't do that is because I wasn't sure if I'm allowed to await futures that come from grpcio in a task spawned by the tokio stuff (that was partly why I asked questions earlier).

Anyways, it seems like there are fewer questions that arise if I just make it an `std::thread`, and the worker doesn't really need parallelism or more than one thread. So I wasn't sure what was gained by making it a tokio task instead.

Note also that, most of our rust web services right now, including consensus, fog ingest, fog view, etc. all spawn background threads this way and communicate either using queues, or sometimes just std lib mutexes. So I guess I just felt very comfortable making the background worker an std thread.

The other thing was, by making it a tokio task, it's possible that the worker thread can be starved if there are a lot requests that come in, because all the OS threads could be assigned to the rocket task I guess. Maybe it doesn't really matter. But it seemed a little easier to think about in what circumstances the worker could get blocked if I don't make it use the same thread pool as the HTTP server.

---

I'm sure that there are things that can be improved about the worker thread if it is async, the main one being, instead of sleeping a fixed amount of time, you could do some kind of select over the one-shot receivers and the timer, so that it wakes up immediately if there might be something to do. You could also do that using an std condition variable I think but select is better for various reasons. But I'm not sure if having the worker thread wake up immediately in those scenarios matters. And just using `std::thread` this way seemed simpler to me.

---

Anyways if you see a major benefit to changing it to `tokio::Task` then I'm happy to change this, but I'm curious if in your mind this is more than substituting `std::thread::spawn` with `tokio::task` and `std::thread::sleep` with `tokio::time::sleep.await`.

I don't see that the Streams stuff is relevant here -- if it became necessary to use those APIs, we could turn the mpsc queue endpoint into a stream and then use it that way: https://docs.rs/tokio-stream/0.1.8/tokio_stream/wrappers/struct.ReceiverStream.html

Reply:
Also thank you for doing this, this is an *excellent* feature

Reply:
Yeah I'm good with the way this is. It works (and is impressive engineering..) and it solves a pretty significant problem in our release cycles, so it should go in.

As to the engineering philosophy bit, I think there's a bit of an art to mixing threads and executors. Std::thread + an async executor is pretty common in practice I think and a valid approach. Sync code can pass as effectively async through the mspc channels between threads because the channels act as a stream that that queue up done tasks as is done here.  But there are other ways and I think my main point is let's huddle more as a team on async design in the design phase of things like this as I think that's a path towards having a team mastery of the async paradigm.

Thanks for building this!